src/spatial/delaunay.c

```diff
@@ -0,0 +1,260 @@
+     * become necessary if linking to an external Qhull. */
+    QHULL_LIB_CHECK
+
+    /* Initializes qh, sets qh->qhull_command. */
+    qh_init_A(qh, NULL, NULL, NULL, 0, NULL);
+    IGRAPH_FINALLY(destroy_qhull, qh);
+
+    exitcode = setjmp(qh->errexit);
+    if (!exitcode) {
+        qh->NOerrexit = False;
+
+        /* qh_option() does not change the operation of Qhull. It simply records options to be
+         * output with error messages. Here we manually keep it in sync with the settings below. */
+        qh_option(qh, "delaunay  Qz-infinity-point Q3-no-merge-vertices", NULL, NULL);
+
+        qh->PROJECTdelaunay = True; // project points to parabola to calculate delaunay triangulation
+        qh->DELAUNAY = True; // 'd'
+        qh->ATinfinity = True; // 'Qz', required for cocircular points
+        qh->MERGEvertices = False; // 'Q3', do not merge identical vertices
+
+        qh_initflags(qh, qh->qhull_command);
+        qh_init_B(qh, qhull_points, numpoints, dim, ismalloc); // read points and project them to parabola.
+        qh_qhull(qh); // do the triangulation
+        qh_triangulate(qh); // this guarantees that everything is simplicial
+
+        igraph_vector_int_t simplex;
+        IGRAPH_VECTOR_INT_INIT_FINALLY(&simplex, dim + 1); // a simplex in n dimensions has n+1 incident vertices.
+
+        facetT *facet; // required for FORALLfacets
+        vertexT *vertex, **vertexp; // required for FOREACHvertex_
+
+        FORALLfacets {
+            if (!facet->upperdelaunay) {
+                igraph_integer_t curr_vert = 0;
+                FOREACHvertex_(facet->vertices) {
+                    VECTOR(simplex)[curr_vert++] = qh_pointid(qh, vertex->point);
+                }
+                IGRAPH_CHECK(add_clique(edges, &simplex));
+            }
+        }
+
+        IGRAPH_CHECK(igraph_i_simplify_edge_list(edges, false, false, false));
+
+        /* Check if there are any points/vertices that do not appear in the edge list.
+         * This happens when there are duplicate points, as Qhull ignores one of them.
+         * We raise an error when there are duplicates. */
+        igraph_integer_t edges_size = igraph_vector_int_size(edges);
+        igraph_bitset_t in_edge_list;
+
+        IGRAPH_BITSET_INIT_FINALLY(&in_edge_list, numpoints);
+        for (igraph_integer_t i = 0; i < edges_size; i++) {
+            IGRAPH_BIT_SET(in_edge_list, VECTOR(*edges)[i]);
+        }
+        if (igraph_bitset_is_any_zero(&in_edge_list)) {
+            IGRAPH_ERROR("Duplicate points for Delaunay triangulation.", IGRAPH_EINVAL);
+        }
+
+        igraph_bitset_destroy(&in_edge_list);
+        igraph_vector_int_destroy(&simplex);
+        destroy_qhull(qh);
+        igraph_free(qhull_points);
+        IGRAPH_FINALLY_CLEAN(4);
+    } else {
+        switch (qh->last_errcode) {
+        /* TODO: More specific error descriptions for common Qhull errors. */
+        default:
+            /* TODO: Report Qhull error text? */
```

Comment:
That would be nice but not necessary if it is hard to extract (e.g. if we need to duplicate Qhull's error strings here).